File: backend/app/db/models/autonomous.py
from __future__ import annotations

from datetime import datetime, date, timezone
from typing import Optional

from sqlalchemy import Boolean, DateTime, Date, Float, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class AutonomousAction(Base):
    __tablename__ = "autonomous_actions"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    lab: Mapped[str] = mapped_column(String, nullable=False)
    strategy_id: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    action_type: Mapped[str] = mapped_column(String, nullable=False)
    asset: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    params: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    rationale: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    result: Mapped[str] = mapped_column(String, nullable=False, default="success")
    outcome_value: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False, index=True,
        default=lambda: datetime.now(timezone.utc),
    )


class AutonomousGuardrail(Base):
    __tablename__ = "autonomous_guardrails"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int] = mapped_column(Integer, nullable=False, index=True, unique=True)
    daily_loss_limit_pct: Mapped[float] = mapped_column(Float, nullable=False, default=3.0)
    weekly_loss_limit_pct: Mapped[float] = mapped_column(Float, nullable=False, default=7.0)
    max_consecutive_losses: Mapped[int] = mapped_column(Integer, nullable=False, default=5)
    max_drawdown_pct: Mapped[float] = mapped_column(Float, nullable=False, default=25.0)
    max_position_concentration: Mapped[float] = mapped_column(Float, nullable=False, default=15.0)
    max_open_positions: Mapped[int] = mapped_column(Integer, nullable=False, default=20)
    autonomous_paused: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    paused_reason: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    paused_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )


class AutonomousDigest(Base):
    __tablename__ = "autonomous_digests"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    digest_date: Mapped[date] = mapped_column(Date, nullable=False)
    tldr: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    highlights: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    strategy_spotlight: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    signals_fired: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    paper_buys: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    paper_sells: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    best_action: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    worst_action: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    tomorrow_preview: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    market_regime_note: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    full_narrative: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    portfolio_delta: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )

File: frontend/src/api/autonomous.ts
import client from "./client";

export interface AutonomousStatus {
  engine_running: boolean;
  strategies_active: number;
  total_assets_monitored: number;
  open_positions: number;
  last_action_at: string | null;
  last_action_summary: string | null;
  guardrail_status: "ok" | "tripped" | "paused";
  autonomous_paused: boolean;
  market_status: "open" | "pre-market" | "after-hours" | "closed";
}

export interface AutonomousAction {
  id: number;
  lab: string;
  strategy_id: string;
  action_type: string;
  asset: string;
  params: Record<string, unknown>;
  rationale: string;
  result: "success" | "skipped" | "failed";
  outcome_value: number | null;
  created_at: string;
}

export interface AutonomousGuardrail {
  daily_loss_limit_pct: number;
  weekly_loss_limit_pct: number;
  max_consecutive_losses: number;
  max_open_positions: number;
  autonomous_paused: boolean;
  paused_reason: string | null;
}

export interface AutonomousStats {
  signals_fired: number;
  paper_buys: number;
  paper_sells: number;
  stop_hits: number;
  target_hits: number;
  pnl_24h: number;
  best_action: { asset: string; pnl: number } | null;
  worst_action: { asset: string; pnl: number } | null;
  by_lab: Record<string, number>;
}

export interface AutonomousDigest {
  id: number;
  digest_date: string;
  tldr: string;
  highlights: string[];
  strategy_spotlight: string;
  signals_fired: number;
  paper_buys: number;
  paper_sells: number;
  best_action: string | null;
  tomorrow_preview: string;
  market_regime_note: string;
  full_narrative: string;
  portfolio_delta: number;
}

export const getStatus = () =>
  client.get<AutonomousStatus>("/autonomous/status").then(r => r.data);

export const getActionsFeed = () =>
  client.get<any>("/autonomous/actions/feed").then(r => {
    const raw = r.data;
    if (Array.isArray(raw)) return { actions: raw as AutonomousAction[] };
    return raw as { actions: AutonomousAction[] };
  });

export const getStats24h = () =>
  client.get<AutonomousStats>("/autonomous/stats/24h").then(r => r.data);

export const getGuardrails = () =>
  client.get<AutonomousGuardrail>("/autonomous/guardrails").then(r => r.data);

export const pauseAutonomous = (reason?: string) =>
  client.post<AutonomousGuardrail>("/autonomous/pause", { reason }).then(r => r.data);

export const resumeAutonomous = () =>
  client.post<AutonomousGuardrail>("/autonomous/resume").then(r => r.data);

export const updateGuardrails = (data: Partial<AutonomousGuardrail>) =>
  client.put<AutonomousGuardrail>("/autonomous/guardrails", data).then(r => r.data);

export const getLatestDigest = () =>
  client.get<AutonomousDigest | null>("/autonomous/digest/latest").then(r => r.data);

export const getDigestHistory = () =>
  client.get<{ digests: AutonomousDigest[] }>("/autonomous/digest/history").then(r => r.data);
